Solutions/Marain.Workflow.Abstractions/Marain/Workflows/WorkflowEngine.cs

```diff
@@ -124,7 +128,19 @@ private async Task ProcessInstanceAsync(IWorkflowTrigger trigger, string instanc
             {
                 if (item?.IsDirty == true)
                 {
-                    await this.workflowInstanceStore.UpsertWorkflowInstanceAsync(item, partitionKey).ConfigureAwait(false);
+                    try
+                    {
+                        await Task.WhenAll(
+                            this.workflowInstanceChangeLog.RecordWorkflowInstanceChangeAsync(trigger, item, partitionKey),
+                            this.workflowInstanceStore.UpsertWorkflowInstanceAsync(item, partitionKey)).ConfigureAwait(false);
```

Comment:
Rather than doing them in parallel, might it make more sense to persist the change entry first, then the instance itself?

We can more easily recover from a successful change log entry followed by a failed instance update, so by only attempting to persist the instance once we've written the change log, we make things easier to clean up if something goes wrong and the engine faults.

Reply:
See the ADR for why we are not doing this right now. Primarily

1. This is not an ACID-supporting transaction log by any means
2. Executing in parallel has minimal impact on overall performance
3. It doesn't blow up until they've all had a go, so if our log store went down, but not our instance store (or vice-versa) we'd still get a record of the state in one or the other which could be used to recover.